README.txt
    README for PDL::VectorValued

ABSTRACT
    PDL::VectorValued - Assorted PDL utilities treating vectors as values

REQUIREMENTS
    *   PDL

        Tested version(s) 2.4.2, 2.4.3, 2.4.7_001, 2.4.9_015, 2.4.10, 2.019,
        2.039

DESCRIPTION
    PDL::VectorValued provides some generalizations of builtin PDL functions
    to higher order PDLs which treat vectors in the source PDLs as "data
    values".

BUILDING
    Build this module as you would any perl module, by doing something akin
    to the following:

     gzip -dc distname-XX.YY.tar.gz | tar -xof -
     cd distname-XX.YY/
     perl Makefile.PL
     make
     make test                                     # optional
     make install

    See perlmodinstall(1) for details.

ACKNOWLEDGEMENTS
    *   Perl by Larry Wall

    *   PDL by Karl Glazebrook, Tuomas J. Lukka, Christian Soeller, and
        others.

    *   Code for rlevec() and rldvec() derived from the PDL builtin
        functions rle() and rld() in $PDL_SRC_ROOT/Basic/Slices/slices.pd

    *   Code for vv_qsortvec() copied nearly verbatim from the builtin PDL
        functions in $PDL_SRC_ROOT/Basic/Ufunc/ufunc.pd, with Chris
        Marshall's "uniqsortvec" patch. Code for vv_qsortveci() based on the
        same.

AUTHOR
    Bryan Jurish <moocow@cpan.org>

COPYRIGHT
    Copyright (c) 2007-2021, Bryan Jurish. All rights reserved.

    This package is free software, and entirely without warranty. You may
    redistribute it and/or modify it under the same terms as Perl itself.



ChangeLog
##-*- Mode: Change-Log; coding: utf-8; -*-
##
## Change log for perl distribution PDL::VectorValued

v1.0.19 Wed, 18 May 2022 14:41:44 +0200 moocow
	* add vv_ prefix to all VectorValued::Utils functions
	* add conditional aliases in VectorValued.pm BEGIN block for github issue #5
	  - import %VV_IMPORT symbols from PDL core if available
	  - export vv_FOO as PDL::FOO otherwise (compatibility)

v1.0.18 Wed, 16 Mar 2022 20:23:23 +0100 moocow
	* removed redundant mv(1,0)->...->mv(0,1) from setops trimming expressions

v1.0.17 Sun, 13 Mar 2022 17:43:18 +0100 moocow
	* overhaul set operations implicit threading, part 2
	  - use RedoDimsCode for setting default dimensions (adapted from PDL::Primitive::append)
	  - drop klunky (& sometimes wrong) dimension-wrangling stuff in setops PMCode
	  - add PP code to explicitly zero out unpopulated portions of set-op result piddles ($c)
	    + might be worth allowing user to specify "missing" value here (a la PDL::CCS)
	* added more tests from https://github.com/moocow-the-bovine/PDL-VectorValued/issues/4

v1.0.16 Sat, 12 Mar 2022 15:48:13 +0100 moocow
	* overhaul set operations implicit threading/broadcasting dimensions and empty results
	  - should hopefully fix https://github.com/moocow-the-bovine/PDL-VectorValued/issues/4

v1.0.15 Sun, 20 Feb 2022 13:40:45 +0100 moocow
	* fix for PDL v2.075
	  - PDL::null() is no longer a valid input PDL, even for conversions

v1.0.14 Fri, 12 Nov 2021 20:05:32 +0100 moocow
	* explicitly use PDL::Types from VectorValued/Dev; fixes https://github.com/moocow-the-bovine/PDL-VectorValued/issues/1

v1.0.13 Mon, 20 Sep 2021 17:30:18 +0200 moocow
	* cleaned up stale acknowledgements in docs

v1.0.12 Sun, 19 Sep 2021 21:24:50 +0200 moocow
	* removed stale vv_qsortvec and vv_qsortveci cruft
	  - original motivation (bug in PDL-2.4.3) is long since fixed
	  - vv_* methods are now just deprecated glob-aliases for the corresponding builtin PDL core functions

v1.0.11 Thu, 16 Sep 2021 20:42:46 +0200 moocow
	* replaced crufty $COMP(__X_size) with $SIZE(X) in Utils/utils.pd
	  - at suggestion of Ed J, for upcoming PDL release
	* added github repo to CPAN metadata

v1.0.10 Wed, 07 Apr 2021 09:52:45 +0200 moocow
	* applied Zaki Mughal's patch for PDL native complex types in PDL > 2.028
	  - https://github.com/zmughal/PDL-VectorValued/compare/skip-complex-types
	  - fixes RT bug #134962

v1.0.9 Fri, 18 May 2018 09:44:25 +0200 moocow
	* added README.rpod to MANIFEST
	* only include README.txt in "realclean" targets if README.rpod is present

v1.0.8 Tue, 15 May 2018 08:55:00 +0200 moocow
	* fixed some typos in POD documentation (RT ticket #125314)

v1.0.7 Tue, 06 Jun 2017 10:21:58 +0200 moocow
	* added 'use lib "."' to Makefile.PL (RT bug #121661)

v1.0.6 Mon, 06 Jun 2016 11:29:23 +0200 moocow
	* win32/NaN fixes for tests (RT bug #115076)

v1.0.5 Thu, 17 Dec 2015 10:02:25 +0100 moocow
	* fixed wrongly succeeding bogus tests with unary ok()
	  - tests now use Test::More and re-factored common test subroutines
	* pdlmaker.plm doesn't distribute generated PM files any more (PDL now does this for us)
	* added vector-cosine similarity vv_vcos()

v1.0.4 Mon, 23 Nov 2015 12:10:32 +0100 moocow
	* shared $VERSION now via perl-reversion script from Perl::Version (debian libperl-version-perl)
	  as suggested by Corion in thread http://www.perlmonks.org/?node_id=1148116
	* may or may not help with weird errors building PDL::Ngrams with stale PDL::VectorValued,
	  - new version of PDL::VectorValued will probably mask that bug anyways

v1.0.3 Tue, 04 Aug 2015 12:19:25 +0200 moocow
	* enumvec() regression fix

v1.0.2 Wed, 29 Jul 2015 10:05:34 +0200 moocow
	* enumvecg() fix

v1.0.1 Tue, 28 Jul 2015 16:51:18 +0200 moocow
	* added enumvecg(): global vector id-enumerator

v1.0.0 Mon, 27 Jul 2015 10:06:35 +0200 moocow
	* explicit 3-component versioning for more transparent version.pm compatiblity
	* added rleseq(), rldseq() for run-length encoding subsequence vectors

v0.09002 Thu, 09 Apr 2015 10:31:50 +0200 moocow
	* tweaked bootstrap logic snafu with PDL::VectorValued::Dev including PDL::VectorValued::Version
	  even though ::Dev is needed at module build time when ::Version is not yet present
	  - this seemed to be causing a lot of UNKNOWN results on cpantesters

v0.09001 Wed, 08 Apr 2015 10:58:04 +0200 moocow
	* added $CMPVAL() and $LB() macros
	* added v_{union,intersect,setdiff} set operations for flat, sorted, unique-valued PDLs
	  - new implementation is much faster than builtin PDL::Primitive::intersect() [via PDL::Primitive::setops()]
	  - for large-ish intersection (NA=2973, NB=221013) v_intersect($a,qsort($b)) is 2x faster than setops($a,'AND',$b),
	    and v_intersect($a,$b) with all data pre-sorted $b is 42x faster than setops($a,'AND',$b)

v0.08001 2014-11-05  moocow
	* fix annoying PAUSE messages about incompatible version numbers
	* improved handling of empty pdls

v0.07002 Wed, 25 Sep 2013 10:16:54 +0200 moocow
	* PREREQ_PM{PDL}=CONFIGURE_REQUIRES{PDL}=0, for CPAN-testers friendliness
	* added support for PDL v2.007 PDL_Indx type (64-bit indices)
	* changed version convention X.BBCCC to jive with both pp_setversion() and version->parse()

v0.06 2012-11-06  moocow
	* added enumvec(): enumerate repetitions in a sorted vector list

v0.05 2012-01-02  moocow
	* pdlmaker.plm version: for CPAN-friendly docs

v0.0401 2011-12-20  moocow
	* added 'generic' pdlmaker.plm : cpan-friendly pdl module hacks
	* still not playing too nicely with embedded pd->pm conversions
	  (e.g. Utils/utils.pd -> Utils/utils.pm)

v0.04 Wed, 30 Mar 2011 15:12:32 +0200 (moocow)
	+ renamed qsortveci() to vv_qsortveci to avoid conflicts with pdl-2.4.7 (debian squeeze)

v0.03 Mon, 16 Apr 2007 09:58:49 +0200
	+ added @PDL::VectorValued::Dev::MACROS

v0.02 Tue, 10 Apr 2007 14:45:00 +0200 (moocow)
	+ added setops: vv_union, vv_intersect, vv_setdiff

v0.01 Fri, 06 Apr 2007 11:41:00 +0200 (moocow)
	+ initial version


VectorValued.pm
## $Id$
##
## File: PDL::VectorValued.pm
## Author: Bryan Jurish <moocow@cpan.org>
## Description: Vector utilities for PDL: perl side only
##======================================================================

package PDL::VectorValued;
use strict;

##======================================================================
## Export hacks
use PDL;
use PDL::Exporter;
use PDL::VectorValued::Utils;
our @ISA = qw(PDL::Exporter);

our (@EXPORT_OK);
BEGIN {
  ##--------------------------------------------------------------------
  ## Conditional bindings for PDL > v2.079
  ##  + see https://github.com/moocow-the-bovine/PDL-VectorValued/issues/5

  ## @VV_SYMBOLS : exportable symbols (vv_FOO)
  my @VV_SYMBOLS =
    (
     (@PDL::VectorValued::Utils::EXPORT_OK), ##-- inherited
     qw(vv_uniqvec),
     qw(vv_rleND vv_rldND),
     qw(vv_indx),
    );

  # %VV_IMPORT: import these from PDL core if available (PDL > v2.079)
  my %VV_IMPORT = (
		   vv_rlevec => {vv=>'rlevec', p=>PDL->can('rlevec')},
		   vv_rldvec => {vv=>'rldvec', p=>PDL->can('rldvec')},
		   vv_rleseq => {vv=>'rleseq', p=>PDL->can('rleseq')},
		   vv_rldseq => {vv=>'rldseq', p=>PDL->can('rldseq')},
		   vv_enumvec => {vv=>'enumvec', p=>PDL->can('enumvec')},
		   vv_enumvecg => {vv=>'enumvecg', p=>PDL->can('enumvecg')},
		   vv_vsearchvec => {vv=>'vsearchvec', p=>PDL->can('vsearchvec')},
		   vv_cmpvec => {vv=>'cmpvec', p=>PDL->can('cmpvec')},
		   vv_union => {vv=>'vv_union', p=>PDL->can('unionvec')},
		   vv_intersect => {vv=>'vv_intersect', p=>PDL->can('intersectvec')},
		   vv_setdiff => {vv=>'vv_setdiff', p=>PDL->can('setdiffvec')},
		   v_union => {vv=>'v_union', p=>PDL->can('union_sorted')},
		   v_intersect => {vv=>'v_intersect', p=>PDL->can('intersect_sorted')},
		   v_setdiff => {vv=>'v_setdiff', p=>PDL->can('setdiff_sorted')},
		   vv_rleND => {vv=>'rleND', p=>undef},
		   vv_rldND => {vv=>'rldND', p=>undef},
		   #vv_indx => {vv=>'vv_indx', p=>PDL->can('indx')}, # DEBUG
		  );


  @EXPORT_OK = @VV_SYMBOLS;
  foreach my $vv_sym (@VV_SYMBOLS) {
    no strict 'refs';

    if ($VV_IMPORT{$vv_sym} && defined($VV_IMPORT{$vv_sym}{p})) {
      # function lives in PDL core: import it here, and clobber $vv_sym here (but not in VV::Utils)
      *$vv_sym = *{$VV_IMPORT{$vv_sym}{vv}} = $VV_IMPORT{$vv_sym}{p};
    }
    elsif ($VV_IMPORT{$vv_sym}) {
      # $sym is defined here as "vv_$sym" : bind it here & in PDL namespace
      my $sym = $VV_IMPORT{$vv_sym}{vv};
      ${PDL::}{$sym} = *$sym = *$vv_sym;

      # ... and make it exportable
      push(@EXPORT_OK, $sym);
    }
  }
}

our %EXPORT_TAGS =
  (
   Func => [@EXPORT_OK],               ##-- respect PDL conventions (hopefully)
  );

## VERSION was formerly set by PDL::VectorValued::Version, now use perl-reversion from Perl::Version instead
our $VERSION = '1.0.19';

##======================================================================
## pod: header
=pod

=head1 NAME

PDL::VectorValued - Utilities for vector-valued PDLs

=head1 SYNOPSIS

 use PDL;
 use PDL::VectorValued;

 ##---------------------------------------------------------------------
 ## ... stuff happens

=cut

##======================================================================
## Description
=pod

=head1 DESCRIPTION

PDL::VectorValued provides generalizations of some elementary PDL
functions to higher-order PDLs which treat vectors as "data values".

=cut


##======================================================================
## pod: Aliases
=pod

=head1 ALIASES

To facilitate incorporation of selected vector-valued functions into
the PDL core, the PDL:PP-, XS-, C-, and perl-level functions defined by
this module in the C<PDL::VectorValued::Utils> package namespace
all carry a C<vv_> prefix as of PDL::VectorValued v1.0.19
Prior to v1.0.19, many of these functions (e.g. C<cmpvec()>)
were defined by this module without a C<vv_> prefix.

For PDL::VectorValued E<gt>= v1.0.19 and PDL E<gt> v2.079, most vector-valued
functions are expected to be defined in the PDL core.  For such "moving" functions
C<FUNC>, the PDL core implementations will be imported into the C<PDL::VectorValued>
namespace as both C<FUNC> and C<vv_FUNC>, clobbering any local implementation
from the C<PDL::VectorValued::Utils> namespace.  Local implementations
C<vv_FUNC> which were previously defined and exported as C<FUNC>
or for which no binding for C<*PDL::FUNC> exists will be bound to
both C<*PDL::VectorValued::FUNC> and C<*PDL::FUNC>, and exported by default,
for backwards-compatibility.

Functions expected to move to the PDL core are:


=over 4

=item *

New code should use C<FUNC()> or C<PDL::FUNC()>.

=item *

Backwards-compatible code can use C<FUNC()> or C<PDL::VectorValued::FUNC()>.

=item *

Direct use of C<PDL::VectorValued::Utils::vv_FUNC()> is deprecated.

=item *

Direct use of C<PDL::VectorValued::Utils::FUNC()> is likely broken as of
PDL::VectorValued v1.0.19.


=back

=cut


##======================================================================
## pod: Functions
=pod

=head1 FUNCTIONS

=cut

##----------------------------------------------------------------------
## vv_uniqvec

=pod

=head2 vv_uniqvec

=for sig

  Signature: (v(N,M); [o]vu(N,MU))

=for ref

Drop-in replacement for broken uniqvec() which uses vv_qsortvec().
Otherwise copied from PDL::Primitive::primitive.pd.

See also: PDL::VectorValued::Utils::vv_qsortvec, PDL::Primitive::uniqvec.

=cut

*PDL::vv_uniqvec = \&vv_uniqvec;
sub vv_uniqvec {
  my($pdl) = shift;

  # slice is not cheap but uniqvec isn't either -- shouldn't cost too much.
  return $pdl if($pdl->nelem == 0 || $pdl->ndims <2 || $pdl->slice("(0)")->nelem < 2); 

  my $srt = $pdl->mv(0,-1)->
    clump($pdl->ndims - 1)->
      mv(-1,0)->vv_qsortvec-> ##-- moo: Tue, 24 Apr 2007 17:17:39 +0200: use vv_qsortvec
	mv(0,-1);

  $srt=$srt->dice($srt->mv(0,-1)->ngoodover->which) if ($PDL::Bad::Status && $srt->badflag);
  ##use dice instead of nslice since qsortvec might be packing the badvals to the front of
  #the array instead of the end like the docs say. If that is the case and it gets fixed,
  #it won't bust uniqvec. DAL 14-March 2006
  my $uniq = ($srt != $srt->rotate(-1)) -> mv(0,-1) -> orover->which;

  return $uniq->nelem==0 ? 
	$srt->slice("0,:")->mv(0,-1) :
	$srt->dice($uniq)->mv(0,-1);
}


##======================================================================
## Run-Length Encoding/Decoding: n-dimensionl
=pod

=head1 Higher-Order Run-Length Encoding and Decoding

The following functions generalize the builtin PDL functions rle() and rld()
for higher-order "values".

See also:
PDL::VectorValued::Utils::vv_rlevec(), PDL::VectorValued::Utils::vv_rldvec().

=cut

##----------------------------------------------------------------------
## rleND()
=pod

=head2 vv_rleND

=for sig

  Signature: (data(@vdims,N); int [o]counts(N); [o]elts(@vdims,N))

=for ref

Run-length encode a set of (sorted) n-dimensional values.

Generalization of rle() and vv_rlevec():
given set of values $data, generate a vector $counts with the number of occurrences of each element
(where an "element" is a matrix of dimensions @vdims ocurring as a sequential run over the
final dimension in $data), and a set of vectors $elts containing the elements which begin a run.
Really just a wrapper for clump() and rlevec().

See also: PDL::Slices::rle, PDL::Ngrams::VectorValued::Utils::vv_rlevec.

=cut

*PDL::vv_rleND = \&vv_rleND;
sub rleND {
  my $data   = shift;
  my @vdimsN = $data->dims;

  ##-- construct output pdls
  my $counts = $#_ >= 0 ? $_[0] : zeroes(long, $vdimsN[$#vdimsN]);
  my $elts   = $#_ >= 1 ? $_[1] : zeroes($data->type, @vdimsN);

  ##-- guts: call rlevec()
  vv_rlevec($data->clump($#vdimsN), $counts, $elts->clump($#vdimsN));

  return ($counts,$elts);
}

##----------------------------------------------------------------------
## rldND()
=pod

=head2 vv_rldND

=for sig

  Signature: (int counts(N); elts(@vdims,N); [o]data(@vdims,N);)

=for ref

Run-length decode a set of (sorted) n-dimensional values.

Generalization of rld() and rldvec():
given a vector $counts() of the number of occurrences of each @vdims-dimensioned element,
and a set $elts() of @vdims-dimensioned elements, run-length decode to $data().

Really just a wrapper for clump() and rldvec().

See also: PDL::Slices::rld, PDL::VectorValued::Utils::rldvec

=cut

*PDL::vv_rldND = \&vv_rldND;
sub vv_rldND {
  my ($counts,$elts) = (shift,shift);
  my @vdimsN        = $elts->dims;

  ##-- construct output pdl
  my ($data);
  if ($#_ >= 0) { $data = $_[0]; }
  else {
    my $size      = $counts->sumover->max; ##-- get maximum size for Nth-dimension for small encodings
    my @countdims = $counts->dims;
    shift(@countdims);
    $data         = zeroes($elts->type, @vdimsN, @countdims);
  }

  ##-- guts: call rldvec()
  vv_rldvec($counts, $elts->clump($#vdimsN), $data->clump($#vdimsN));

  return $data;
}

##======================================================================
## pod: Functions: datatype utilities
=pod

=head1 Datatype Utilities

=cut

##----------------------------------------------------------------------
## vv_indx()
=pod

=head2 vv_indx

=for sig

  Signature: vv_indx()

=for ref

Returns PDL::Type subclass used for indices.
If built with PDL E<lt> v2.007, this should return C<PDL::long>, otherwise C<PDL::indx>.

=cut

sub vv_indx {
  return defined(&PDL::indx) ? PDL::indx(@_) : PDL::long(@_);
}


##======================================================================
## pod: Functions: low-level
=pod

=head2 Low-Level Functions

Some additional low-level functions are provided in the
PDL::VectorValued::Utils
package.
See L<PDL::VectorValued::Utils> for details.

=cut



1; ##-- make perl happy

##======================================================================
## pod: Footer
=pod

=head1 ACKNOWLEDGEMENTS

perl by Larry Wall.

=head1 AUTHOR

Bryan Jurish E<lt>moocow@cpan.orgE<gt>

PDL by Karl Glazebrook, Tuomas J. Lukka, Christian Soeller, and others.

=head1 COPYRIGHT

Copyright (c) 2007-2022, Bryan Jurish.  All rights reserved.

This package is free software.  You may redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

perl(1), PDL(3perl), PDL::VectorValued::Utils(3perl)

=cut
